src/task.c

```diff
@@ -632,109 +627,53 @@ task_enqueue(task_system *ts,
       tq->head = tq->tail = new_task;
    }
 
-   __sync_fetch_and_add(&group->current_outstanding_tasks, 1);
+   __sync_fetch_and_add(&group->current_waiting_tasks, 1);
 
    if (group->use_stats) {
       new_task->enqueue_time = platform_get_timestamp();
    }
-   if (group->current_outstanding_tasks > group->max_outstanding_tasks) {
-      group->max_outstanding_tasks = group->current_outstanding_tasks;
-   }
-   return task_unlock_task_queue(group);
-}
-
-/*
- * task_perform_all() - Perform all tasks queued with the task system.
- * Returns as soon as it finds the queue is empty.
- */
-void
-task_perform_all(task_system *ts)
-{
-   if (!ts->use_bg_threads) {
-      platform_status rc;
-      do {
-         rc = task_perform_one(ts);
-         debug_assert(SUCCESS(rc) || STATUS_IS_EQ(rc, STATUS_TIMEDOUT));
-      } while (STATUS_IS_NE(rc, STATUS_TIMEDOUT));
-
-   } else {
-      // wait for bg threads to finish running all the queued up tasks.
-      for (task_type type = TASK_TYPE_FIRST; type != NUM_TASK_TYPES; type++) {
-         task_group *group = &ts->group[type];
-         while (group->current_outstanding_tasks != 0) {
-            platform_sleep(USEC_TO_NSEC(100000)); // 100 msec.
-         }
+   if (group->use_stats) {
+      const threadid tid = platform_get_tid();
+      if (group->current_waiting_tasks
+          > group->stats[tid].max_outstanding_tasks) {
+         group->stats[tid].max_outstanding_tasks = group->current_waiting_tasks;
       }
    }
+   platform_condvar_signal(&group->cv);
+   return platform_condvar_unlock(&group->cv);
 }
 
-static void
-task_queue_unlock(void *arg)
-{
-   task_group *group = (task_group *)arg;
-   platform_assert(!task_system_use_bg_threads(group->ts));
-   platform_mutex_unlock(&group->fg.mutex);
-}
-
-static inline platform_status
-task_group_perform_one(task_group *group)
+static platform_status
+task_group_perform_one(task_group *group, uint64 min_backlog)
 {
-   platform_assert(!task_system_use_bg_threads(group->ts));
    platform_status rc;
    task           *assigned_task = NULL;
-   task_queue     *tq            = &group->tq;
-   if (group->current_outstanding_tasks == 0) {
+   if (group->current_waiting_tasks <= min_backlog) {
       return STATUS_TIMEDOUT;
    }
 
-   platform_thread_cleanup_push(task_queue_unlock, group);
-   rc = platform_mutex_lock(&group->fg.mutex);
+   platform_thread_cleanup_push((void (*)(void *))platform_condvar_unlock,
+                                &group->cv);
+   rc = platform_condvar_lock(&group->cv);
```

Comment:
If this fails, then on L659    we do: `   goto out;`

Then, in the `out:` block, on L 669  we may attempt to do: `platform_condvar_unlock(&group->cv);`
That may also complain.

It's difficult to repro this situation, but from code-walk through there may be a latent problem. 

I think some subtle rearrangement of the placement of code under `out: ` label is needed.

I could not immediately think of what rework to suggest, as it involved handling properly `assigned_task` and `__sync_fetch_and_add()` calls.

Reply:
Good catch.

It turns out there are no cancellation points in the critical section, so we don't need to push a thread cleanup function at all.

Cleaned up based on that.